apache/spark
# [SPARK-17283][Core] Cancel job in RDD.take() as soon as enough output is received

## What changes were proposed in this pull request?

This patch introduces a new internal implementation of `RDD.take()` which halts the Spark job as soon as enough rows have been received. This can offer substantial performance improvements for interactive queries. For example, consider

```
sc.parallelize(1 to 100000, 100000).filter(_ > 10000).take(1)
```

In the current `take()` implementation, this will end up computing almost every partition of the RDD even though only the first 10% of partitions must be computed to return the desired output. With this patch's changes, Spark will cancel the running job as soon as enough output has been received, greatly improving responsiveness for limit queries with highly-selective filters.

Although the changes here are implemented in Spark core, they also benefit Spark SQL; the following query also observes similar speedups:

```
sc.parallelize(1 to 100000, 100000).toDF.filter("value > 10000").take(1)
```

**Caveat:** the performance optimizations implemented here do not apply to `AsyncRDDActions.takeAsync()`. The `takeAsync()` codepath is fairly complicated and, to the best of my knowledge, is not very widely used, so I'd like to postpone any changes to it.
### Implementation details

**Task result buffering:** this new implementation of `take()` returns exactly the same answers as the old implementation. The old implementation first collected the per-partition results from all partitions and then processed them in ascending order of partition id. In order to match this behavior, the new implementation buffers task outputs in order to process them in order.

**Scheduler changes:** the existing `FutureAction.cancel()` method causes the Spark scheduler to mark the task as failed, which doesn't make sense in this context. To address this, this patch modifies the DAGScheduler to allow jobs to be cancelled while still marking them as successful. This is a slightly tricky change that will require discussion and more tests (to be added here).
### Feature flags

I have made this new implementation the default but have preserved the old `take()` implementations behind feature-flags.

In Spark Core, the old `take()` implementation can be restored by setting `spark.driver.adaptiveTakeEnabled=false` in SparkConf.

In Spark SQL, the old implementation can be used by setting `spark.sql.adaptiveTake.enabled=false` in SQLConf.
## How was this patch tested?

**Manual tests:**
- Ensure that the console progress bar disappears after the job halts early:
  
  ```
   sc.parallelize(1 to 100000, 100000).toDF.filter("value > 10000").rdd.map(_ => Thread.sleep(1)).take(1)
  ```
- Ensure that the driver thread (the thread which calls `take()`) is interrupted / unblocked in case the user cancels the active `take()` job via the web UI or via a job group.

**Automated tests:**
- Ensure that a `take()` command whose Spark tasks fail with exceptions still causes the right exceptions to be surfaced to the driver.
- DAGSchedulerSuite test for the new "cancel without failing job" scheduler functionality.
